src/http/http_rpc_context.h

```diff
@@ -0,0 +1,355 @@
+    {
+      return std::nullopt;
+    }
+
+    ctx.set_method(remaining_path);
+    return actor;
+  }
+
+  class HttpRpcContext : public enclave::RpcContext
+  {
+  private:
+    http_method verb;
+    std::vector<uint8_t> request_body;
+    std::string path = {};
+
+    http::HeaderMap request_headers;
+
+    mutable jsonrpc::Pack body_packing = jsonrpc::Pack::Text;
+
+  public:
+    HttpRpcContext(
+      const enclave::SessionContext& s,
+      http_method verb_,
+      const std::string_view& path_,
+      const std::string_view& query_,
+      const http::HeaderMap& headers_,
+      const std::vector<uint8_t>& body_,
+      const std::vector<uint8_t>& raw_request_ = {},
+      const std::vector<uint8_t>& raw_pbft_ = {}) :
+      RpcContext(s, raw_request_, raw_pbft_),
+      verb(verb_),
+      path(path_),
+      request_body(body_)
+    {
+      // Build a canonical serialization of this request. If the request is
+      // signed, then all unsigned headers must be removed
+      request_headers = headers_;
+
+      const auto auth_it = request_headers.find(http::headers::AUTHORIZATION);
+      if (auth_it != request_headers.end())
+      {
+        std::string_view authz_header = auth_it->second;
+
+        auto parsed_sign_params =
+          http::HttpSignatureVerifier::parse_signature_params(authz_header);
+
+        if (!parsed_sign_params.has_value())
+        {
+          throw std::logic_error(fmt::format(
+            "Unable to parse signature params from: {}", authz_header));
+        }
+
+        // Keep all signed headers, and the auth header containing the signature
+        // itself
+        auto& signed_headers = parsed_sign_params->signed_headers;
+        signed_headers.emplace_back(http::headers::AUTHORIZATION);
+
+        for (const auto& required_header :
+             {http::headers::DIGEST, http::headers::CONTENT_LENGTH})
+        {
+          if (
+            std::find(
+              signed_headers.begin(), signed_headers.end(), required_header) ==
+            signed_headers.end())
+          {
+            throw std::logic_error(fmt::format(
+              "HTTP authorization header must sign header '{}'",
+              required_header));
+          }
+        }
+
+        auto it = request_headers.begin();
+        while (it != request_headers.end())
+        {
+          if (
+            std::find(
+              signed_headers.begin(), signed_headers.end(), it->first) ==
+            signed_headers.end())
+          {
+            it = request_headers.erase(it);
+          }
+          else
+          {
+            ++it;
+          }
+        }
+      }
+
+      const auto canonical_request_header = fmt::format(
+        "{} {} HTTP/1.1\r\n"
+        "{}"
+        "\r\n",
+        http_method_str(verb),
+        fmt::format("{}{}", path_, query_),
+        http::get_header_string(request_headers));
+
+      raw_request.resize(canonical_request_header.size() + request_body.size());
```

Comment:
This seems like a stupid way of building the `raw_request`, but its the best I've found. The `http_parser` API doesn't give us an easy way to track the _start_ and _end_ of a parsed message in the bytestream we gave it, and we'd have to buffer separately to support split messages. Technically we could produce this later, on-demand, but I think that merely delays the work.